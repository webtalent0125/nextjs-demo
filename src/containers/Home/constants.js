export const GET_SHOWCASES = 'app/Example/GET_SHOWCASES';
