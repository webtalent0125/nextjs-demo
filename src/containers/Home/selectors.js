import { createSelector } from 'reselect';

import { initialState } from './reducer';

const selectShowcasesDomain = state => state.showcases || initialState;

const selectShowcases = () => createSelector(selectShowcasesDomain, subState => subState);

export { selectShowcases, selectShowcasesDomain };
