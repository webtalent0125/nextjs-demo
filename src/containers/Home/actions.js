import { createRoutine } from 'redux-saga-routines';

import { GET_SHOWCASES } from './constants';

export const getShowcases = createRoutine(GET_SHOWCASES);
