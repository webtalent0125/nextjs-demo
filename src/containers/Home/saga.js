import { takeLatest, put, call } from 'redux-saga/effects';

import request from 'utils/request';

import { getShowcases } from './actions';

export function* getShowcasesRequest() {
  const BASE_URL = 'https://us-central1-react-next-boilerplate-cda8b.cloudfunctions.net/getShowcasesData';

  try {
    yield put(getShowcases.request());

    const showcasesdata = yield call(request, BASE_URL);

    yield put(getShowcases.success(showcasesdata));
  } catch (err) {
    yield put(getShowcases.failure(err));
  } finally {
    yield put(getShowcases.fulfill());
  }
}

export default function* dataShowcases() {
  yield takeLatest(getShowcases.TRIGGER, getShowcasesRequest);
}
