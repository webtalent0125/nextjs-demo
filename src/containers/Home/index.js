import React, { memo } from 'react';
import PropTypes from 'prop-types';

import { compose } from 'redux';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import { useInjectReducer } from 'utils/inject-reducer';
import { useInjectSaga } from 'utils/inject-saga';

import Layout from 'components/Layout';
import Features from 'components/Features';
import Showcases from 'components/Showcases';

import saga from './saga';
import reducer from './reducer';
import { getShowcases } from './actions';
import { selectShowcases } from './selectors';

export function Home({ getShowcases, showcasesData }) {
  useInjectSaga({ key: 'showcases', saga });
  useInjectReducer({ key: 'showcases', reducer });

  return (
    <Layout>
      <Features />

      <Showcases onGetShowcases={getShowcases} data={showcasesData} />
    </Layout>
  );
}

const mapStateToProps = createStructuredSelector({
  showcasesData: selectShowcases(),
});

export function mapDispatchToProps(dispatch) {
  return { getShowcases: () => dispatch(getShowcases()) };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

Home.propTypes = {
  showcasesData: PropTypes.object,
  getShowcases: PropTypes.func,
};

export default compose(withConnect, memo)(Home);
