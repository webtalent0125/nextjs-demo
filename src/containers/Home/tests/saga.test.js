import { put, takeLatest } from 'redux-saga/effects';

import { getShowcases } from 'containers/Home/actions';

import dataShowcases, { getShowcasesRequest } from '../saga';

describe('getShowcasesRequest Saga', () => {
  let mainSaga;

  beforeEach(() => {
    mainSaga = dataShowcases();
  });

  it('Should get list of showcases TRIGGER', () => {
    const takeLatestDescriptor = mainSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(getShowcases.TRIGGER, getShowcasesRequest));
  });

  describe('getShowcases actions', () => {
    let getShowcasesGenerate;

    const response = [
      {
        title: 'First title',
      },
      {
        title: 'Second title',
      },
    ];

    beforeEach(() => {
      getShowcasesGenerate = getShowcasesRequest();

      const selectDescriptor = getShowcasesGenerate.next().value;
      expect(selectDescriptor).toMatchSnapshot();
    });

    it('Should getShowcasesRequest be success', () => {
      getShowcasesGenerate.next(response);
      const putShowcasesSuccess = getShowcasesGenerate.next(response).value;
      getShowcasesGenerate.next(response);

      expect(putShowcasesSuccess).toEqual(put(getShowcases.success(response)));
    });

    it('Should getShowcasesRequest be failure', () => {
      const response = new Error('Some error');
      const putShowcasesFailure = getShowcasesGenerate.throw(response).value;

      expect(putShowcasesFailure).toEqual(put(getShowcases.failure(response)));
    });
  });
});
