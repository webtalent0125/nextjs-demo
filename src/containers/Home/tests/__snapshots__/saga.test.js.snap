// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getShowcasesRequest Saga getShowcases actions Should getShowcasesRequest be failure 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "type": "app/Example/GET_SHOWCASES/REQUEST",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`getShowcasesRequest Saga getShowcases actions Should getShowcasesRequest be success 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "type": "app/Example/GET_SHOWCASES/REQUEST",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;
